### PR TITLE
6567 syscollector shared library

### LINKED ISSUE #6567: change Syscollector to shared library
## Description
With the fact of simplifying the distribution and being backward compatible with older versions of cmake (2.6.4 is the minimum required).

Syscollector must be compiled as a shared library and separately the .o files must be removed from the link, which will be replaced with the .dll / .so respectively.

- [x] Change code/make files
- [x] Run RTR
- [x] PR

### PR DESCRIPTION
**Pull request by Dwordcito**
|Related issue|
|---|
|Closes #6567 |

# Changes included
- Provider normalization.
- Folder structure change.
- RTR tool changes.
- Change to shared libraries.
- Fix nlohmann initialization list issue
- Compilation Issues
- Remove old syscollector implementation

image
image
image
image

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -1968,11 +1970,11 @@ win32_ui_o := $(win32_ui_c:.c=.o)
...
 
-win32/ossec-agent.exe: win32/icon.o win32/win_agent.o win32/win_service.o win32/win_utils.o ${syscollector_o} ${syscheck_o} ${rootcheck_o} $(filter-out wazuh_modules/main.o, ${wmodulesd_o}) $(filter-out client-agent/main.o, $(filter-out client-agent/agentd.o, $(filter-out client-agent/event-forward.o, ${client_agent_o}))) $(filter-out logcollector/main.o, ${os_logcollector_o}) ${os_execd_o} monitord/rotate_log.o monitord/compress_log.o wazuh_modules/syscollector/syscollector_win_ext.dll
-	${OSSEC_CXXBIN} -DARGV0=\"ossec-agent\" -DOSSECHIDS ${OSSEC_LDFLAGS} $^ ${OSSEC_LIBS} ${SHARED_MODULES_LDFLAGS} -o $@
+win32/ossec-agent.exe: win32/icon.o win32/win_agent.o win32/win_service.o win32/win_utils.o ${syscheck_o} ${rootcheck_o} $(filter-out wazuh_modules/main.o, ${wmodulesd_o}) $(filter-out client-agent/main.o, $(filter-out client-agent/agentd.o, $(filter-out client-agent/event-forward.o, ${client_agent_o}))) $(filter-out logcollector/main.o, ${os_logcollector_o}) ${os_execd_o} monitord/rotate_log.o monitord/compress_log.o
+	${OSSEC_CCBIN} -DARGV0=\"ossec-agent\" -DOSSECHIDS ${OSSEC_LDFLAGS} $^ ${OSSEC_LIBS} ${SYSINFO_LIB} ${SYSCOLLECTOR_LIB} ${DBSYNC_LIB} -o $@
```
Comment (konangc87) [inline src/Makefile:1974]: isn't rsync library missing here?

File: src/Makefile [hunk 2]
```diff
@@ -1968,11 +1970,11 @@ win32_ui_o := $(win32_ui_c:.c=.o)
...
-	${OSSEC_CXXBIN} -DARGV0=\"ossec-agent\" -DOSSECHIDS -DEVENTCHANNEL_SUPPORT ${OSSEC_LDFLAGS} $^ ${OSSEC_LIBS} ${SHARED_MODULES_LDFLAGS} -o $@
+win32/ossec-agent-eventchannel.exe: win32/icon.o win32/win_agent.o win32/win_service.o win32/win_utils.o $(filter-out syscheckd/main-event.o, ${syscheck_eventchannel_o}) ${rootcheck_o} $(filter-out wazuh_modules/main.o, ${wmodulesd_o}) $(filter-out client-agent/main.o, $(filter-out client-agent/agentd.o, $(filter-out client-agent/event-forward.o, ${client_agent_o}))) $(filter-out logcollector/main-event.o, ${os_logcollector_eventchannel_o}) ${os_execd_o} monitord/rotate_log.o monitord/compress_log.o
+	${OSSEC_CCBIN} -DARGV0=\"ossec-agent\" -DOSSECHIDS -DEVENTCHANNEL_SUPPORT ${OSSEC_LDFLAGS} $^ ${OSSEC_LIBS} ${SYSINFO_LIB} ${SYSCOLLECTOR_LIB} ${DBSYNC_LIB} -o $@
```
Comment (konangc87) [inline src/Makefile:1977]: ditto

File: src/data_provider/tests/sysInfoNetworkLinux/sysInfoNetworkLinux_test.cpp [hunk 3]
```diff
@@ -29,7 +29,7 @@ class SysInfoNetworkLinuxWrapperMock: public INetworkInterfaceWrapper
     virtual ~SysInfoNetworkLinuxWrapperMock() = default;
     MOCK_METHOD(int, family, (), (const override));
     MOCK_METHOD(std::string, name, (), (const override));
-    MOCK_METHOD(std::string, adapter, (), (const override));
+    MOCK_METHOD(std::string, description, (), (const override));
```
Comment (konangc87) [inline src/data_provider/tests/sysInfoNetworkLinux/sysInfoNetworkLinux_test.cpp:32]: remove trailing spaces

File: src/data_provider/tests/sysInfoNetworkLinux/sysInfoNetworkLinux_test.cpp [hunk 4]
```diff
@@ -38,13 +38,14 @@ class SysInfoNetworkLinuxWrapperMock: public INetworkInterfaceWrapper
     MOCK_METHOD(std::string, broadcastV6, (), (const override));
     MOCK_METHOD(std::string, gateway, (), (const override));
     MOCK_METHOD(std::string, metrics, (), (const override));
-    MOCK_METHOD(std::string, metricsV6, (), (const override));
+    MOCK_METHOD(std::string, metricsV6, (), (const override));
```
Comment (konangc87) [inline src/data_provider/tests/sysInfoNetworkLinux/sysInfoNetworkLinux_test.cpp:41]: ditto

File: src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp [hunk 5]
```diff
@@ -28,7 +28,7 @@ class SysInfoNetworkWindowsWrapperMock: public INetworkInterfaceWrapper
     virtual ~SysInfoNetworkWindowsWrapperMock() = default;
     MOCK_METHOD(int, family, (), (const override));
     MOCK_METHOD(std::string, name, (), (const override));
-    MOCK_METHOD(std::string, adapter, (), (const override));
+    MOCK_METHOD(std::string, description, (), (const override));
```
Comment (konangc87) [inline src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp:31]: ditto

File: src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp [hunk 6]
```diff
@@ -37,7 +37,7 @@ class SysInfoNetworkWindowsWrapperMock: public INetworkInterfaceWrapper
     MOCK_METHOD(std::string, broadcastV6, (), (const override));
     MOCK_METHOD(std::string, gateway, (), (const override));
     MOCK_METHOD(std::string, metrics, (), (const override));
-    MOCK_METHOD(std::string, metricsV6, (), (const override));
+    MOCK_METHOD(std::string, metricsV6, (), (const override));
```
Comment (konangc87) [inline src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp:40]: ditto

File: src/data_provider/tests/sysInfoNetworkBSD/sysInfoNetworkBSD_test.cpp [hunk 7]
```diff
@@ -111,7 +110,7 @@ TEST_F(SysInfoNetworkBSDTest, Test_AF_LINK)
     EXPECT_EQ("eth01",ifaddr.at("name").get_ref<const std::string&>());
     EXPECT_EQ("ethernet",ifaddr.at("type").get_ref<const std::string&>());
     EXPECT_EQ("up",ifaddr.at("state").get_ref<const std::string&>());
-    EXPECT_EQ("00:A0:C9:14:C8:29",ifaddr.at("mac").get_ref<const std::string&>());
+    EXPECT_EQ("00:A0:C9:14:C8:29",ifaddr.at("MAC").get_ref<const std::string&>());
```
Comment (tinachoB) [inline src/data_provider/tests/sysInfoNetworkBSD/sysInfoNetworkBSD_test.cpp:113]: In `networkInterfaceBSD.cpp` file, all values are lower case. This key (and the one changed in line 123) should be updated back to lower case.

File: src/shared_modules/dbsync/src/sqlite/sqlite_dbengine.cpp [hunk 8]
```diff
@@ -833,8 +833,7 @@ bool SQLiteDBEngine::getPKListLeftOnly(const std::string& t1,
 }
 
 std::string SQLiteDBEngine::buildDeleteBulkDataSqlQuery(const std::string& table,
-                                                        const std::vector<std::string>& primaryKeyList,
-                                                        const nlohmann::json& jsData)
+                                                        const std::vector<std::string>& primaryKeyList)
```
Comment (konangc87) [inline src/shared_modules/dbsync/src/sqlite/sqlite_dbengine.cpp:836]: is this right? I remember we needed to add this in order to support the delete without all the primary keys.

File: src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp (hunk 5 above)
Comment (tinachoB) [inline src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp:31]: In `inetworkWrapper.h` interface, there is no field called "description". This has been changed by "adapter" in previous PR changes.

File: src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp (hunk 6 above)
Comment (tinachoB) [inline src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp:40]: Please, remove extra line endings

File: src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp [hunk 9]
```diff
@@ -98,7 +98,7 @@ TEST_F(SysInfoNetworkWindowsTest, Test_IPV4)
     EXPECT_EQ(netmask, networkInfo.at("IPv4").at("netmask").get_ref<const std::string&>());
     EXPECT_EQ(broadcast, networkInfo.at("IPv4").at("broadcast").get_ref<const std::string&>());
     EXPECT_EQ(dhcp, networkInfo.at("IPv4").at("dhcp").get_ref<const std::string&>());
-    EXPECT_EQ(metrics, networkInfo.at("IPv4").at("metric").get_ref<const std::string&>());
+    EXPECT_EQ(metrics, networkInfo.at("IPv4").at("metrics").get_ref<const std::string&>());
```
Comment (tinachoB) [inline src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp:101]: metrics -> metric

File: src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp [hunk 10]
```diff
@@ -121,7 +121,7 @@ TEST_F(SysInfoNetworkWindowsTest, Test_IPV6)
     EXPECT_EQ(netmask, networkInfo.at("IPv6").at("netmask").get_ref<const std::string&>());
     EXPECT_EQ(broadcast, networkInfo.at("IPv6").at("broadcast").get_ref<const std::string&>());
     EXPECT_EQ(dhcp, networkInfo.at("IPv6").at("dhcp").get_ref<const std::string&>());
-    EXPECT_EQ(metrics, networkInfo.at("IPv6").at("metric").get_ref<const std::string&>());
+    EXPECT_EQ(metrics, networkInfo.at("IPv6").at("metrics").get_ref<const std::string&>());
```
Comment (tinachoB) [inline src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp:124]: ditto

File: src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp [hunk 11]
```diff
@@ -148,7 +148,7 @@ TEST_F(SysInfoNetworkWindowsTest, Test_COMMON_DATA)
     EXPECT_EQ(name, networkInfo.at("name").get_ref<const std::string&>());
     EXPECT_EQ(type, networkInfo.at("type").get_ref<const std::string&>());
     EXPECT_EQ(state, networkInfo.at("state").get_ref<const std::string&>());
-    EXPECT_EQ(MAC, networkInfo.at("mac").get_ref<const std::string&>());
+    EXPECT_EQ(MAC, networkInfo.at("MAC").get_ref<const std::string&>());
```
Comment (tinachoB) [inline src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp:151]: MAC -> mac

File: src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp [hunk 12]
```diff
@@ -159,6 +159,6 @@ TEST_F(SysInfoNetworkWindowsTest, Test_COMMON_DATA)
     EXPECT_EQ(7, networkInfo.at("tx_dropped").get<int32_t>());
     EXPECT_EQ(6, networkInfo.at("rx_dropped").get<int32_t>());
 
-    EXPECT_EQ(mtu, networkInfo.at("mtu").get_ref<const std::string&>());
+    EXPECT_EQ(mtu, networkInfo.at("MTU").get_ref<const std::string&>());
```
Comment (tinachoB) [inline src/data_provider/tests/sysInfoNetworkWindows/sysInfoNetworkWindows_test.cpp:162]: MTU -> mtu

File: src/shared_modules/dbsync/tests/interface/dbsync_test.cpp [hunk 13]
```diff
@@ -1282,7 +1282,7 @@ TEST_F(DBSyncTest, deleteSingleDataByCompoundPK)
 
     EXPECT_EQ(0, dbsync_sync_row(handle, jsInitialData.get(), callbackData));  // Expect an insert event
     EXPECT_EQ(0, dbsync_delete_rows(handle, jsSingleDeletion.get()));
-    EXPECT_EQ(0, dbsync_delete_rows(handle, jsMissingPKPID.get()));
+    EXPECT_NE(0, dbsync_delete_rows(handle, jsMissingPKPID.get()));
```
Comment (konangc87) [inline src/shared_modules/dbsync/tests/interface/dbsync_test.cpp:1285]: =O this definitely looks wrong.

File: src/shared_modules/rsync/src/rsyncImplementation.cpp [hunk 14]
```diff
@@ -56,13 +56,10 @@ void RSyncImplementation::startRSync(const RSYNC_HANDLE handle,
     const auto& jsStartParamsTable  { jsStartParams.at("table")              };
     const auto& firstQuery          { jsStartParams.find("first_query")      };
     const auto& lastQuery           { jsStartParams.find("last_query")       };
-
+
```
Comment (konangc87) [inline src/shared_modules/rsync/src/rsyncImplementation.cpp:59]: remove trailing space

File: src/wazuh_modules/syscollector/include/syscollector.h [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+ * November 15, 2020.
+ *
+ * This program is free software; you can redistribute it
+ * and/or modify it under the terms of the GNU General Public
+ * License (version 2) as published by the FSF - Free Software
+ * Foundation.
+ */
+
+
+#ifndef _SYSCOLLECTOR_INFO_H
```
Comment (tinachoB) [inline src/wazuh_modules/syscollector/include/syscollector.h:13]: _SYSCOLLECTOR_INFO_H -> _SYSCOLLECTOR_H

File: src/wazuh_modules/syscollector/include/syscollector.hpp [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+ * Wazuh SysCollector
+ * Copyright (C) 2015-2020, Wazuh Inc.
+ * October 8, 2020.
+ *
+ * This program is free software; you can redistribute it
+ * and/or modify it under the terms of the GNU General Public
+ * License (version 2) as published by the FSF - Free Software
+ * Foundation.
+ */
+#ifndef _SYSCOLLECTOR_IMP_H
```
Comment (tinachoB) [inline src/wazuh_modules/syscollector/include/syscollector.hpp:11]: _SYSCOLLECTOR_IMP_H -> _SYSCOLLECTOR_HPP

File: src/wazuh_modules/syscollector/src/syscollectorImp.cpp [hunk 17]
```diff
@@ -338,30 +306,8 @@ static void updateAndNotifyChanges(const DBSYNC_HANDLE handle,
...
-    {
-        const auto daysToHours{m_intervalValue * 24ul};
-        ret = !m_cv.wait_for(lock, std::chrono::hours{daysToHours}, [&](){return m_running;});
-    }
-    else
-    {
-        ret = !m_cv.wait_for(lock, std::chrono::hours{1}, [&](){return m_running;});
-    }
-    return ret;
+    return !m_cv.wait_for(lock, std::chrono::seconds{m_intervalValue}, [&](){return m_running;});;
```
Comment (konangc87) [inline src/wazuh_modules/syscollector/src/syscollectorImp.cpp:310]: remove extra ;

**Review by tinachoB — CHANGES_REQUESTED** [review]

File: src/wazuh_modules/wm_syscollector.c [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+                               sys->flags.osinfo,
+                               sys->flags.netinfo,
+                               sys->flags.programinfo,
+                               sys->flags.portsinfo,
+                               sys->flags.allports,
+                               sys->flags.procinfo,
+                               sys->flags.hotfixinfo);
+    }
+    
+    return 0;
```
Comment (konangc87) [inline src/wazuh_modules/wm_syscollector.c:81]: return an error code if the library can't be loaded.

File: src/wazuh_modules/wm_syscollector.h [hunk 19]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Wazuh Module for System inventory
+ * Copyright (C) 2015-2020, Wazuh Inc.
+ * March 9, 2017.
```
Comment (konangc87) [inline src/wazuh_modules/wm_syscollector.h:4]: wrong date

**Review by konangc87 — CHANGES_REQUESTED** [review]

**Review by tinachoB — APPROVED** [review]

**Review by konangc87 — APPROVED** [review]

**Review by tinachoB — APPROVED** [review]

**Pull request merged**